src/Microsoft.AspNetCore.OData/EnableNestedPathsAttribute.cs

```diff
@@ -0,0 +1,150 @@
+            {
+                // This could be a SingleResult, which has the property Queryable.
+                // But it could be a SingleResult() or SingleResult<T>. Sort by number of parameters
+                // on the property and get the one with the most parameters.
+                PropertyInfo propInfo = responseContent.Value.GetType().GetProperties()
+                    .OrderBy(p => p.GetIndexParameters().Count())
+                    .Where(p => p.Name.Equals("Queryable"))
+                    .LastOrDefault();
+
+                result = propInfo.GetValue(singleResult) as IQueryable;
+            }
+
+            if (result != null)
+            {
+                ODataPath path = actionExecutedContext.HttpContext.ODataFeature().Path;
+                IEdmModel model = actionExecutedContext.HttpContext.Request.GetModel();
+
+                var queryBuilder = new ODataPathQueryBuilder(result, model, path);
+                ODataPathQueryResult transformedResult = queryBuilder.BuildQuery();
+
+                if (transformedResult == null)
+                {
+                    // TODO: is this the best way to return 404
+                    actionExecutedContext.Result = new NotFoundObjectResult(null);
+                }
+                else if (path.EdmType.TypeKind == EdmTypeKind.Collection || transformedResult.HasCountSegment)
+                {
+                    responseContent.Value = transformedResult.Result;
+                }
+                else
+                {
+                    Type elementType = transformedResult.Result.ElementType;
+
+                    // If we return the IQueryable as the result, then the response will be returned as a collection
+                    // so we have to return a single result. We can either return the materialized single result using SingleOrDefault,
+                    // which will run the query against the underlying data source. Or we can wrap the IQueryable result with
+                    // a SingleResult<T>.
+                    // If the action has [EnableQuery], then we return a SingleResult<T> which allows query options to be
+                    // applied to the IQueryable. If we returned the materialized object, then query options like $expand
+                    // would return null because navigation properties are not fetched from relational databases by default.
+                    // If the action does not have [EnableQuery], then we run the query using SingleOrDefault and return the
+                    // actual object. If we returned a SingleResult<T> in this case, we'd get an error because the serializer
+                    // does not recognize it.
+                    if (actionDescriptor.MethodInfo.GetCustomAttributes<EnableQueryAttribute>().Any())
+                    {
+                        // calls SingleResult.Create<T>(IQueryable<T>)
+                        MethodInfo singleResultCreate = typeof(SingleResult)
```

Comment:
I think you could move the logic for getting `SingleResult` `Create` method to `ExpressionHelperMethods` so its reusable